adaptdl/adaptdl/goodput.py

```diff
@@ -0,0 +1,256 @@
+    "beta_r",   # Retrogression factor of intra-node network time
+    # T_step ~ (T_compute ^ gamma + T_network ^ gamma) ^ (1 / gamma)
+    # Essentially is a p-norm where p = gamma. When p ~ 1 then
+    # T_step ~ T_compute + T_network, indicating no overlap between compute
+    # and network. When p -> infinity then T_step = max(T_compute, T_network),
+    # indicating perfect overlap. We limit gamma to [1, 10] since 10 is close
+    # enough to approximate the max function for our purposes.
+    "gamma",    # Models the degree of overlap between compute and network
+])
+
+
+class GoodputFunction(object):
+
+    def __init__(self, perf_params, grad_params, init_batch_size):
+        self._perf_params = perf_params
+        self._grad_params = grad_params
+        self._init_batch_size = init_batch_size
+
+    def __call__(self, num_nodes, num_replicas, atomic_bsz, accum_steps):
+        return self.evaluate(num_nodes, num_replicas, atomic_bsz, accum_steps)
+
+    def evaluate(self, num_nodes, num_replicas, atomic_bsz, accum_steps):
+        return (self.throughput(num_nodes, num_replicas,
+                                atomic_bsz, accum_steps) *
+                self.efficiency(num_replicas * atomic_bsz * (accum_steps + 1)))
+
+    def throughput(self, num_nodes, num_replicas, atomic_bsz, accum_steps):
+        step_time, _, _ = _predict(self._perf_params, num_nodes,
+                                   num_replicas, atomic_bsz, accum_steps)
+        batch_size = num_replicas * atomic_bsz * (accum_steps + 1)
+        return batch_size / step_time
+
+    def efficiency(self, batch_size):
+        grad_sqr = self._grad_params['norm']
+        grad_var = self._grad_params['var']
+        scale = batch_size * self._init_batch_size
+        denom = grad_var / scale + grad_sqr
+        gain = np.where(denom > 0, (grad_var + grad_sqr) / denom, 1.0)
+        return gain / scale
+
+    def optimize(self, num_nodes, num_replicas, max_batch_size=None,
+                 atomic_bsz_range=None, accumulation=False):
+        assert np.all(np.less_equal(1, num_nodes))
+        assert np.all(np.less_equal(num_nodes, num_replicas))
+        if max_batch_size is None:
+            max_batch_size = self._init_batch_size
+        assert self._init_batch_size <= max_batch_size
+        atomic_bsz_range = atomic_bsz_range or (None, None)
+        min_atomic_bsz = atomic_bsz_range[0] or 1
+        max_atomic_bsz = atomic_bsz_range[1] or max_batch_size
+        # Remember what the output shape/format should be and flatten inputs.
+        output_shape = np.broadcast(num_nodes, num_replicas).shape
+        output_scalar = np.isscalar(num_nodes) or np.isscalar(num_replicas)
+        num_nodes = np.broadcast_to(num_nodes, output_shape).flatten()
+        num_replicas = np.broadcast_to(num_replicas, output_shape).flatten()
+        # Samples 50 different total batch sizes in geometric space.
+        batch_size = np.geomspace(self._init_batch_size, max_batch_size)
```

Comment:
Can't the local batch size be smaller than `self._init_batch_size`? For example, with `self._init_batch_size` = 128, no minimum local bsz, no accumulation and on two replicas, the local batch size could be 64.

Reply:
Yeah it can, defined on the line below.